### PR TITLE
Add CFA model

### LINKED ISSUE #615: Implementing "CFA for Target-Oriented Anomaly Localization"
### Discussed in https://github.com/openvinotoolkit/anomalib/discussions/388

<div type='discussions-op-text'>

<sup>Originally posted by **alexriedel1** June 15, 2022</sup>
Hello,
I was just wondering if someone is already working on an anomalib implementation of https://arxiv.org/pdf/2206.04325.pdf
(https://github.com/sungwool/CFA_for_anomaly_localization)

It apparently scores higher than PatchCore on MVTec without any tricks like ensembling very large models. </div>

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- This PR adds the implementation of the CFA Model
- Fixes #615

## TODO:
- Create the benchmark results.

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/cfa/README.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+### NOTE
+
+When the numbers are produced, early stopping callback (patience: 5) is used. It might be possible to achieve higher-metrics by increasing the patience.
+
+---
+
+### Image-Level AUC
+
+|            | ResNet-18 | Wide ResNet50 |
```
Comment (ashwinvaidya17) [inline anomalib/models/cfa/README.md:37]: Minor issue but we use a horizontal layout in other models.

File: anomalib/models/cfa/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+| Tile       |   0.740   |     0.768     |
+| Toothbrush |   0.609   |     0.627     |
+| Transistor |   0.570   |     0.666     |
+| Wood       |   0.564   |     0.627     |
+| Zipper     |   0.561   |     0.668     |
+| Average    |   0.574   |     0.598     |
+
+### Sample Results
+
+![Sample Result 1](../../../docs/source/images/cfa/results/0.png "Sample Result 1")
```
Comment (ashwinvaidya17) [inline anomalib/models/cfa/README.md:142]: These result images are not in this PR

File: configs/model/cfa.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  gpus: null # Set automatically
+  gradient_clip_val: 0
+  ipus: null
+  limit_predict_batches: 1.0
+  limit_test_batches: 1.0
+  limit_train_batches: 1.0
+  limit_val_batches: 1.0
+  log_every_n_steps: 50
+  log_gpu_memory: null
+  max_epochs: 30
```
Comment (ashwinvaidya17) [inline configs/model/cfa.yaml:86]: Is this what the paper recommends? We have `max_epochs` as 100 for all the models except for those where the paper requires more.

File: anomalib/models/components/filters/blur.py [hunk 4]
```diff
@@ -19,32 +31,37 @@ class GaussianBlur2d(nn.Module):
...
+        kernel_size: Optional[Union[Tuple[int, int], int]] = None,
         normalize: bool = True,
         border_type: str = "reflect",
         padding: str = "same",
     ) -> None:
         """Initialize model, setup kernel etc..
 
         Args:
-            kernel_size (Union[Tuple[int, int], int]): size of the Gaussian kernel to use.
             sigma (Union[Tuple[float, float], float]): standard deviation to use for constructing the Gaussian kernel.
```
Comment (ashwinvaidya17) [inline anomalib/models/components/filters/blur.py:44]: Not part of this PR but should we move these init_args above to make it consistent with other classes?

File: anomalib/models/cfa/config.yaml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  gpus: null # Set automatically
+  gradient_clip_val: 0
+  ipus: null
+  limit_predict_batches: 1.0
+  limit_test_batches: 1.0
+  limit_train_batches: 1.0
+  limit_val_batches: 1.0
+  log_every_n_steps: 50
+  log_gpu_memory: null
+  max_epochs: 30
```
Comment (ashwinvaidya17) [inline anomalib/models/cfa/config.yaml:92]: Same here. Do we need max 30 epochs?

File: anomalib/models/cfa/torch_model.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+from torch.utils.data import DataLoader
+from torchvision.models.feature_extraction import create_feature_extractor
+from tqdm import tqdm
+
+from anomalib.models.components import DynamicBufferModule, GaussianBlur2d
+
+SUPPORTED_BACKBONES = ("vgg19_bn", "resnet18", "wide_resnet50_2", "efficientnet_b5")
+
+
+# TODO: Replace this with the new torchfx feature extractor.
```
Comment (ashwinvaidya17) [inline anomalib/models/cfa/torch_model.py:30]: Maybe we can add https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/models/components/feature_extractors/torchfx.py here and switch to feature extractor when the feature branch is merged.

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the efforts! I have a few minor comments.

File: anomalib/models/cfa/config.yaml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    patience: 5
+    metric: pixel_AUROC
+    mode: max
+  normalization_method: min_max # options: [null, min_max, cdf]
+
+metrics:
+  image:
+    - AUROC
+  pixel:
+    - AUROC
```
Comment (djdameln) [inline anomalib/models/cfa/config.yaml:41]: To be in line with other models it would be good to add both image- and pixel-level F1 here

File: anomalib/models/cfa/config.yaml [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  threshold:
+    image_default: 0
+    pixel_default: 0
+    adaptive: true
```
Comment (djdameln) [inline anomalib/models/cfa/config.yaml:45]:
```suggestion
    adaptive: true
    image_default: null
    pixel_default: null
```

File: anomalib/models/cfa/lightning_model.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    ) -> None:
+        """Backward step for the CFA model.
+
+        Args:
+            loss (Tensor): Loss value.
+            optimizer (Optional[Optimizer]): Optimizer.
+            optimizer_idx (Optional[int]): Optimizer index.
+        """
+        # TODO: Investigate why retain_graph is needed.
+        loss.backward(retain_graph=True)
```
Comment (djdameln) [inline anomalib/models/cfa/lightning_model.py:87]: Is this really needed? Did you observe different performance when just using the default `backward` method?

File: anomalib/models/cfa/torch_model.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+        target_oriented_features = self.layer(patch_features)
+        return target_oriented_features
+
+
+class CoordConv2d(nn.Conv2d):
+    """CoordConv layer as in the paper.
+
+    Link to the paper: https://arxiv.org/abs/1807.03247
+    Link to the PyTorch implementation: https://github.com/walsvid/CoordConv
```
Comment (djdameln) [inline anomalib/models/cfa/torch_model.py:281]: I guess we should include the copyright notice of the original author somewhere (for example at the top of the file)

File: anomalib/models/cfa/torch_model.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+    """Descriptor module."""
+
+    def __init__(self, gamma_d: int, backbone: str) -> None:
+        super().__init__()
+
+        self.backbone = backbone
+        if self.backbone not in SUPPORTED_BACKBONES:
+            raise ValueError(f"Supported backbones are {SUPPORTED_BACKBONES}. Got {self.backbone} instead.")
+
+        # TODO: Automatically infer the number of dims
```
Comment (djdameln) [inline anomalib/models/cfa/torch_model.py:252]: I think we have a utility function for that in `anomalib.models.components.feature_extractors.utils`

File: anomalib/models/cfa/torch_model.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+            features = list(features.values())
+
+        if hasattr(self, "scale") is False:
+            self.scale = features[0].size()[-2:]
+
+        target_features = self.descriptor(features)
+        distance = self.compute_distance(target_features)
+
+        if self.training:
+            output = self.compute_loss(distance)
```
Comment (djdameln) [inline anomalib/models/cfa/torch_model.py:234]: In our other models we leave the loss computation to the lightning model. Probably best to do the same here for consistency. In this case we could define a `CfaLoss` class which is called from the `training_step` method of the lightning model.

File: anomalib/models/cfa/torch_model.py [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+
+        Returns:
+            Tensor: Anomaly map.
+        """
+        anomaly_map = score.mean(dim=1, keepdim=True)
+        anomaly_map = F.interpolate(anomaly_map, size=self.input_size, mode="bilinear", align_corners=False)
+
+        gaussian_blur = GaussianBlur2d(sigma=4).to(score.device)
+        anomaly_map = gaussian_blur(anomaly_map)  # pylint: disable=not-callable
+        return anomaly_map
```
Comment (djdameln) [inline anomalib/models/cfa/torch_model.py:205]: Should this be wrapped into a `AnomalyMapGenerator` class to match the structure used by other Anomalib models?

File: anomalib/models/components/filters/blur.py [hunk 14]
```diff
@@ -9,6 +9,18 @@
...
+def compute_kernel_size(sigma_val: float) -> int:
+    """Compute kernel size from sigma value.
+
+    Args:
+        sigma_val (float): Sigma value.
+
+    Returns:
+        int: Kernel size.
+    """
+    return 2 * int(4.0 * sigma_val + 0.5) + 1
```
Comment (djdameln) [inline anomalib/models/components/filters/blur.py:21]: We use the same computation in Padim ([here](https://github.com/openvinotoolkit/anomalib/blob/main/anomalib/models/padim/anomaly_map.py#L27)), and maybe other models as well. Would be good to use this function there as well.

**Review by djdameln — COMMENTED** [review]
I have a few comments, all minor

File: anomalib/models/cfa/config.yaml (hunk 5 above)
Comment (samet-akcay) [reply anomalib/models/cfa/config.yaml:95]: As discussed, the original implementation chose 30 epochs.

File: anomalib/models/cfa/lightning_model.py (hunk 9 above)
Comment (samet-akcay) [reply anomalib/models/cfa/lightning_model.py:114]: For some reason, the training does not work without this. We could investigate this in the future.

File: anomalib/models/cfa/torch_model.py (hunk 10 above)
Comment (samet-akcay) [reply anomalib/models/cfa/torch_model.py:235]: Done

File: configs/model/cfa.yaml (hunk 3 above)
Comment (samet-akcay) [reply configs/model/cfa.yaml:86]: Same as above.
